core/src/main/java/org/apache/calcite/sql/type/ArrayElementOperandTypeChecker.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql.type;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperator;
+
+import com.google.common.collect.ImmutableList;
+
+import static org.apache.calcite.sql.type.NonNullableAccessors.getComponentTypeOrThrow;
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * Parameter type-checking strategy where types must be Array and Array element type.
+ */
+public class ArrayElementOperandTypeChecker implements SqlOperandTypeChecker {
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public boolean checkOperandTypes(
+      SqlCallBinding callBinding,
+      boolean throwOnFailure) {
+    final SqlNode op0 = callBinding.operand(0);
+    if (!OperandTypes.ARRAY.checkSingleOperandType(
+        callBinding,
+        op0,
+        0,
+        throwOnFailure)) {
+      return false;
+    }
+
+    RelDataType arrayComponentType =
+        getComponentTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));
+    final SqlNode op1 = callBinding.operand(1);
+    RelDataType aryType1 = SqlTypeUtil.deriveType(callBinding, op1);
```

Comment:
Why not use `getComponentTypeOrThrow`?

Reply:
i am using it  RelDataType arrayComponentType =
        getComponentTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0));

Reply:
Why `op1` not need?
```    
RelDataType biggest =
        callBinding.getTypeFactory().leastRestrictive(
            ImmutableList.of(
                getComponentTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op0)),
                getComponentTypeOrThrow(SqlTypeUtil.deriveType(callBinding, op1))));
```
Here is code in MultisetOperandTypeChecker.
https://github.com/apache/calcite/blob/3e95a13b8c5d715298f469051f792286cea704fa/core/src/main/java/org/apache/calcite/sql/type/MultisetOperandTypeChecker.java#L66

Reply:
op1 it not array type, it is element type. for example array_contains(a[1, 2, 3], 1) 

and why we should  find biggest type , for exmplae array_contains(a[1, 2, 3], 'a')
array type is integer not null array not null -> element type is integer not null
arg1 type is char[1] not null, we can not find  biggest type , so it will throw exception
```
    RelDataType biggest =
        callBinding.getTypeFactory().leastRestrictive(
            ImmutableList.of(arrayComponentType, aryType1));
```